Report

sebastianbagya opened the pull request:
Merging report to dev

sebastianbagya committed:
Merge remote-tracking branch 'refs/remotes/origin/development' into report

sebastianbagya committed:
Finishing Query

sebastianbagya committed:
Report Service + Controller

sebastianbagya committed:
Finalize Report

riencus commented on src/main/java/com/blibli/future/service/ReportServiceImpl.java:
```diff
@@ -0,0 +1,156 @@
+	@Override
+	public List<SingleReportVo> ReportParse(List<Object[]> objects) {
+		if(objects!=null)
+		{
```
Can you explain briefly what this class do to the object? I don't understand what does `ReportParse` means.
Also, method name would be better if written in `lowerCamelCase` (`reportParse`)

sebastianbagya replied:
parse object to SingleReportVo

riencus commented on src/main/java/com/blibli/future/service/ReportServiceImpl.java:
```diff
@@ -0,0 +1,156 @@
+	@Override
+	public List<ReportVo> SummaryReport(String dept, LocalDate startDate, LocalDate endDate) {
+		List<ReportVo> reports = employeeRepository.initReport(dept);
+		List<Object[]> daysComingObject = attendanceRepository.countEmployeeAttendance(dept, startDate, endDate);
```
And this one to, what is `SummaryReport`? Does it generate something or... what? Personally, i think method name would be better if using verb, not noun.

sebastianbagya replied:
fullReport

riencus replied:
I'm sorry for not clear enough, I mean something more like `getFullReport()` or `generateFullReport`whichever explains what the method does.

riencus commented on src/main/java/com/blibli/future/vo/ReportVo.java:
```diff
@@ -0,0 +1,243 @@
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class ReportVo implements Serializable{
+	private static final long serialVersionUID = 2006346112358122493L;
```
Is this a response VO? I think it will be easier to identify between request VO and response Vo if this class is renamed to `ReportResponseVO`.
I'll change the other response class if i happen to meet one

sebastianbagya replied:
response

riencus replied:
Which one is better: `ReportVO`or `ReportResponse` or `ReportResponseVO`?
What's your opinion @sebastianbagya @aamesaa ?

aamesaa replied:
I think  `ReportResponseVO` is better, it can describe more

riencus commented on src/main/java/com/blibli/future/vo/SingleReportVo.java:
```diff
@@ -1,16 +1,25 @@
-public class SickEmployeeVo {
+public class SingleReportVo {
 	private String nik;
 	private String fullName;
```
This one to, is it response or request VO?

sebastianbagya replied:
container class

riencus commented on src/main/java/com/blibli/future/repository/EmployeeAbsencePermitRepository.java:
```diff
@@ -11,14 +11,28 @@
 			+ "from EmployeeAbsencePermit eap JOIN Employee e ON eap.employee.nik = e.nik "
 			+ "WHERE e.nameOfDept = (?1) and eap.absencePermit = (?4) and (eap.requestDate between (?2) and (?3)) group by e.nik ORDER BY count(e) DESC")
-	public List<SickEmployeeVo> countSickEmployee(String deparment, LocalDate startDate, LocalDate endDate, AbsencePermit absencePermit, Pageable pageable);
+	public List<SingleReportVo> countSickEmployee(String deparment, LocalDate startDate, LocalDate endDate, AbsencePermit absencePermit, Pageable pageable);
```
Just asking, `AbsencePermit` in the parameter is used for filtering what kind of absence, isn't it?
If that's the case this method is not only counting the "sick" employee right? Since we have other kind of absence permit.
But if this method is really intended to count just the sick employee, is it possible to put "sick" in the query and remove it from the parameter?

sebastianbagya replied:
see below

riencus reviewed (changes requested):
The one I really ask for a change is the naming problem, the other is a question that I need to clarify

P.S. there's some attribute in `EmployeeYearlyLeave` written in Bahasa, did You really name it in Bahasa or you just forgot to change it?

sebastianbagya committed:
fix report

sebastianbagya committed:
Fix Naming Error

riencus reviewed (commented):
There's some question I would like to get the opinion from you both

riencus commented on src/main/java/com/blibli/future/model/YearlyLeave.java:
```diff
@@ -0,0 +1,139 @@
+	private String nik;
+	private LocalDate startDate;
+	private LocalDate endDate;
+	private int hakCuti;
```
here it is @sebastianbagya, the Bahasa one
I'm sorry wrong class

yogiehartanto replied:
employeeLeaveRight could be better

riencus reviewed (commented):
False accuse

aamesaa reviewed (commented):
my issue is same like Ariel. The naming is a little bit confusing for me

riencus committed:
Refactor Attendance Repository

- rename method
- change countEmployeeLateAttendance's parameter order
- change return type to SingleReportVo
- add indentation for readability

riencus committed:
Refactor EmployeeAbsencePermitRepository

- rename method
- change countAbsencePermit's parameter order
- change return type to SingleReportVo
- add indentation for readability

riencus committed:
Refactor EmployeeYearlyLeaveRepository

- rename method
- change return type to SingleReportVo
- add indentation for readability

riencus committed:
Refactor EmployeeSubstitutionLeaveRightRepository

- rename method
- change return type to SingleReportVo
- add indentation for readability

riencus committed:
Make ReportServiceImpl field private

aamesaa commented on src/main/java/com/blibli/future/repository/AttendanceRepository.java:
```diff
@@ -13,4 +14,19 @@
 	public List<Attendance> findByAttendanceKeyDateBetween(LocalDate dateStart, LocalDate dateEnd);
+	@Query(value = "select e.nik, e.full_Name, e.name_Of_Dept, count(atte) from attendance atte JOIN Employee e ON atte.nik = e.nik WHERE e.name_Of_Dept = (?1) and (atte.date between (?2) and (?3)) group by e.nik", nativeQuery = true)
+	public List<Object[]> countEmployeeAttendance(String department, LocalDate startDate, LocalDate endDate);
+	@Query(value = "select e.nik, e.full_Name, e.name_Of_Dept, count(atte) from attendance atte JOIN Employee e ON atte.nik = e.nik WHERE atte.tap_out is null and e.name_Of_Dept = (?1) and (atte.date between (?2) and (?3)) group by e.nik", nativeQuery = true)
```
Bas, why we need to select the `e.full_Name` and`e.name_Of_Dept`?

aamesaa commented on src/main/java/com/blibli/future/repository/EmployeeAbsencePermitRepository.java:
```diff
@@ -11,14 +11,24 @@
+			+ "when start_Date<=(?2) then (end_Date - (?2))+1 "
+			+ "when end_Date>=(?3) then ((?3)-start_Date)+1 "
+			+ "else (end_Date-start_Date)+1 "
+			+ "end as sumDay "
```
Is `sumDay` is the total selected Day ?

riencus committed:
Simplify Constructor

Remove 0,0... from used constructor and create new Constructor with 3 parameter

riencus committed:
Removing public identifier from all method in interface & revert all native query method

riencus committed:
Simplify parseSingleReportVo method and Generate report function

riencus committed:
Rename SingleResponseVo to SubReportVo

riencus committed:
Rename fullReport and parseSingleReport Object name

sebastianbagya committed:
fixing error from refractor

sebastianbagya committed:
Merge pull request #28 from blibli-future/refactor-report

Refactor report
